sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeOuterJoin.scala

```diff
@@ -271,3 +297,189 @@ private class RightOuterIterator(
+      rightRowKey = rightKeyGenerator(rightRow)
+      numRightRows += 1
+      true
+    } else {
+      rightRow = null
+      rightRowKey = null
+      false
+    }
+  }
+
+  /**
+   * Consume rows from both iterators until their keys are different from the provided matchingKey.
+   */
+  private def findMatchingRows(matchingKey: InternalRow): Unit = {
+    leftMatches.clear()
+    rightMatches.clear()
+    leftIndex = 0
+    rightIndex = 0
+
+    while (leftRowKey != null && keyOrdering.compare(leftRowKey, matchingKey) == 0) {
+      leftMatches += leftRow.copy()
+      advancedLeft()
+    }
+    while (rightRowKey != null && keyOrdering.compare(rightRowKey, matchingKey) == 0) {
+      rightMatches += rightRow.copy()
+      advancedRight()
+    }
+
+    if (leftMatches.size <= leftMatched.capacity) {
+      leftMatched.clear()
+    } else {
+      leftMatched = new BitSet(leftMatches.size)
+    }
+    if (rightMatches.size <= rightMatched.capacity) {
+      rightMatched.clear()
+    } else {
+      rightMatched = new BitSet(rightMatches.size)
+    }
+  }
+
+  /**
+   * Scan for next matching rows in buffers of both sides.
+   * @return true if next row(s) are found and false otherwise.
+   */
+  private def scanNextInBuffered(): Boolean = {
+    while (leftIndex < leftMatches.size) {
+      while (rightIndex < rightMatches.size
+        && !boundCondition(joinedRow(leftMatches(leftIndex), rightMatches(rightIndex)))) {
+        rightIndex += 1
+      }
+      if (rightIndex < rightMatches.size) {
+        // matched with condition
+        leftMatched.set(leftIndex)
+        rightMatched.set(rightIndex)
+        rightIndex += 1
+        return true
+      }
+      leftIndex += 1
+      rightIndex = 0
+      if (!leftMatched.get(leftIndex - 1)) {
+        joinedRow(leftMatches(leftIndex - 1), rightNullRow)
+        return true
+      }
+    }
+
+    while (rightIndex < rightMatches.size && rightMatched.get(rightIndex)) {
+      rightIndex += 1
+    }
+    if (rightIndex < rightMatches.size) {
+      joinedRow(leftNullRow, rightMatches(rightIndex))
+      rightIndex += 1
+      true
+    } else {
+      false
+    }
+  }
+
+  // --- Public methods --------------------------------------------------------------------------
+
+  def getJoinedRow(): JoinedRow = joinedRow
+
+  def advanceNext(): Boolean = {
+    // We already buffered some matching rows, use them directly
+    if (leftIndex <= leftMatches.size || rightIndex <= rightMatches.size) {
+      if (scanNextInBuffered()) {
+        return true
+      }
+    }
+
+    // Both left and right iterators have rows and they can be compared
+    if (leftRow != null && (leftRowKey.anyNull || rightRow == null)) {
+      // Only consume row(s) from left iterator
+      joinedRow(leftRow.copy(), rightNullRow)
+      advancedLeft()
+      true
+    } else if (rightRow != null && (rightRowKey.anyNull || leftRow == null)) {
+      // Only consume row(s) from right iterator
```

Comment:
it's always a single row right? Also this comment doesn't really add much value so I think we can just remove it. Same as the one in L445